composites/Plugin/SnippetEditor/serialization.js

```diff
@@ -64,128 +66,240 @@ export function serializeEditor( rawContent ) {
-		const before = content.substr( 0, index );
-		const between = unserializeTag( content.substr( index, tagValue.length ) );
-		const after = content.substr( index + tagValue.length );
-		content = before + between + after;
+	/*
+	 * Do the actual replacement.
+	 *
+	 * We replace `%%replacement_variable%%` with an entity. The entity is already
+	 * created in the content state. So we can refer to it by using
+	 * `contentState.getLastCreatedEntityKey`.
+	 */
+	const newContentState = Modifier.replaceText(
+		contentState,
+		variableTextSelection,
+		variable.title,
+		// No inline style needed.
+		null,
+		contentState.getLastCreatedEntityKey(),
+	);
 
-		// Decrease the offset by twice the length of the circumfix for every index we replace.
-		const offset = index - i * CIRCUMFIX.length * 2;
-		const key = entityRanges.length;
+	// We need to apply the new content state to the editor state.
+	return EditorState.push( editorState, newContentState, "apply-entity" );
+}
 
-		// Create the DraftJS data.
-		const { entityRange, mappedEntity } = unserializeEntity( key, tag.name, offset );
-		entityRanges.push( entityRange );
-		entityMap[ key ] = mappedEntity;
-	}
+/**
+ * Replaces replacement variables (%%tags%%) in an editor state with entities.
+ *
+ * @param {EditorState} editorState          The editor state to find the variables in.
+ * @param {Array}       replacementVariables The available replacement variables, used
+ *                                           to determine the title in the entities.
+ *
+ * @returns {EditorState} The new editor state with entities.
+ */
+export function replaceReplacementVariables( editorState, replacementVariables ) {
+	const contentState = editorState.getCurrentContent();
+	const blockMap = contentState.getBlockMap();
+	let newEditorState = editorState;
 
-	const blocks = [ {
-		entityRanges,
-		text: content,
-	} ];
+	/*
+	 * Because we do this for each block our code will work for multiple blocks even if
+	 * we currently usually only have one block.
+	 */
+	blockMap.forEach( ( block ) => {
+		const { text, key: blockKey } = block;
+		const foundReplacementVariables = findReplacementVariables( text );
 
-	return {
-		blocks,
-		entityMap,
-	};
+		/*
+		 * Offset keeps track of the amount of shifting that occured by replacing
+		 * replacement variables with entities. This makes sure multiple replacement
+		 * variables are replaced correctly.
+		 */
+		let offset = 0;
+
+		foundReplacementVariables.forEach( ( variable ) => {
+			variable = addTitle( variable, replacementVariables );
+			variable = addPositionInformation( variable, offset );
+
+			let selection = newEditorState.getSelection();
+			selection = moveSelectionAfterReplacement( selection, blockKey, variable );
+
+			newEditorState = replaceVariableWithEntity( newEditorState, variable, blockKey );
+			newEditorState = EditorState.acceptSelection( newEditorState, selection );
+
+			offset = offset + variable.delta;
```

Comment:
Where does this delta come from? What is it? Please comment

Reply:
I'm not sure to understand this offset increment. When typing normally, the whole thing runs at every key stroke. That means offset will always be `0` and it's incremented after the call to `addPositionInformation()`. At the following key stroke, it will be `0` again. It works only on first rendering, then this runs on the whole text and when pasting content that contains multiple replacement variables e.g. `bla bla %%post_title%% bla bla %%sep%% bla bla`. Is this the intended usage?